src/util/mpir_hwtopo.c

```diff
@@ -554,3 +554,61 @@ uint64_t MPIR_hwtopo_get_node_mem(void)
 
     return size;
 }
+
+bool MPIR_hwtopo_is_dev_close_by_name(const char *name)
+{
+    bool is_close = false;
+    if (!bindset_is_valid)
+        return is_close;
+#ifdef HAVE_HWLOC
+    MPIR_hwtopo_gid_t gid = MPIR_hwtopo_get_obj_by_name(name);
+    int hwloc_obj_index = HWTOPO_GET_INDEX(gid);
+    int hwloc_obj_depth = HWTOPO_GET_DEPTH(gid);
+    hwloc_obj_t first_non_io =
+        hwloc_get_obj_by_depth(hwloc_topology, hwloc_obj_depth, hwloc_obj_index);
+    /* Determine if PCI device is "close" to process by checking if process's affinity is included
+     * in PCI device's affinity or if PCI device's affinity is included in rank's affinity */
+    is_close = (hwloc_bitmap_isincluded(bindset, first_non_io->cpuset) ||
+                hwloc_bitmap_isincluded(first_non_io->cpuset, bindset));
+#endif
+    return is_close;
+}
+
+#ifdef HAVE_HWLOC
+static hwloc_obj_t get_first_non_io_obj_by_pci(int domain, int bus, int dev, int func)
+{
+    hwloc_obj_t io_device = hwloc_get_pcidev_by_busid(hwloc_topology, domain, bus, dev, func);
+    MPIR_Assert(io_device);
+    hwloc_obj_t first_non_io = hwloc_get_non_io_ancestor_obj(hwloc_topology, io_device);
+    MPIR_Assert(first_non_io);
+    return first_non_io;
+}
+#endif
+
+bool MPIR_hwtopo_is_dev_close_by_pci(int domain, int bus, int dev, int func)
+{
+    bool is_close = false;
+    if (!bindset_is_valid)
+        return is_close;
+#ifdef HAVE_HWLOC
+    hwloc_obj_t first_non_io = get_first_non_io_obj_by_pci(domain, bus, dev, func);
+    /* Determine if PCI device is "close" to process by checking if process's affinity is included
+     * in PCI device's affinity or if PCI device's affinity is included in rank's affinity */
+    is_close = (hwloc_bitmap_isincluded(bindset, first_non_io->cpuset) ||
+                hwloc_bitmap_isincluded(first_non_io->cpuset, bindset));
```

Comment:
The logic is not given or trivial. How about move it to a utility function so we don't repeat the code?

Reply:
I assume by this, you mean the conditionals here and in `MPIR_hwtopo_is_dev_close_by_name` above?

Reply:
Ok. I've added that.